login/service.go

```diff
@@ -139,76 +129,54 @@ func (gh *gitHubOAuth) Perform(ctx *app.AuthorizeLoginContext) error {
 	// First time access, redirect to oauth provider
 
 	// store referer id to state for redirect later
-	fmt.Println("Got Request from: ", referer)
+	log.Println("Got Request from: ", referer)
 	state = uuid.NewV4().String()
 
 	mapLock.Lock()
 	defer mapLock.Unlock()
 
 	stateReferer[state] = referer
 
-	redirectURL := gh.config.AuthCodeURL(state, oauth2.AccessTypeOnline)
+	keycloak.config.RedirectURL = util.AbsoluteURL(ctx.RequestData, "/api/login/authorize")
+
+	redirectURL := keycloak.config.AuthCodeURL(state, oauth2.AccessTypeOnline)
+
 	ctx.ResponseData.Header().Set("Location", redirectURL)
 	return ctx.TemporaryRedirect()
 }
 
-func (gh gitHubOAuth) getUserEmails(ctx context.Context, token *oauth2.Token) ([]ghEmail, error) {
-	client := gh.config.Client(ctx, token)
-	resp, err := client.Get("https://api.github.com/user/emails")
+func (keycloak keycloakOAuthProvider) getUser(ctx context.Context, token *oauth2.Token) (*openIDConnectUser, error) {
+	client := keycloak.config.Client(ctx, token)
+	resp, err := client.Get("http://sso.demo.almighty.io/auth/realms/demo/protocol/openid-connect/userinfo")
```

Comment:
It looks like there is the only one email per user in Keycloak. If we need to store more emails we could use custom attributes.
I didn't change much my original PR. I was planning to address this one and other improvements in separate PRs. Just not to blow up this PR and minimize possible conflicts with other PRs.